Source/JavaScriptCore/wasm/WasmBBQJIT.cpp

```diff
@@ -5927,39 +5985,312 @@ class BBQJIT {
+        auto isWasmCallee = m_jit.branch64(CCallHelpers::Equal, GPRInfo::regT3, CCallHelpers::TrustedImm32(JSValue::WasmTag));
+        CCallHelpers::JumpList doneCases;
+        {
+            // FIXME: Handling precise allocations in WasmB3IRGenerator catch entrypoints might be unnecessary
+            // https://bugs.webkit.org/show_bug.cgi?id=231213
+            auto preciseAllocationCase = m_jit.branchTestPtr(CCallHelpers::NonZero, GPRInfo::regT0, CCallHelpers::TrustedImm32(PreciseAllocation::halfAlignment));
+            m_jit.andPtr(CCallHelpers::TrustedImmPtr(MarkedBlock::blockMask), GPRInfo::regT0);
+            m_jit.loadPtr(CCallHelpers::Address(GPRInfo::regT0, MarkedBlock::offsetOfHeader + MarkedBlock::Header::offsetOfVM()), GPRInfo::regT0);
+            doneCases.append(m_jit.jump());
+
+            preciseAllocationCase.link(&m_jit);
+            m_jit.loadPtr(CCallHelpers::Address(GPRInfo::regT0, PreciseAllocation::offsetOfWeakSet() + WeakSet::offsetOfVM() - PreciseAllocation::headerSize()), GPRInfo::regT0);
+            doneCases.append(m_jit.jump());
+        }
+
+        isWasmCallee.link(&m_jit);
+        m_jit.loadPtr(CCallHelpers::addressFor(CallFrameSlot::codeBlock), GPRInfo::regT0);
+        m_jit.loadPtr(CCallHelpers::Address(GPRInfo::regT0, Instance::offsetOfVM()), GPRInfo::regT0);
+
+        doneCases.link(&m_jit);
+
+        JIT_COMMENT(m_jit, "restore callee saves from vm entry buffer");
+        m_jit.restoreCalleeSavesFromVMEntryFrameCalleeSavesBuffer(GPRInfo::regT0, GPRInfo::regT3);
+
+        m_jit.loadPtr(CCallHelpers::Address(GPRInfo::regT0, VM::callFrameForCatchOffset()), GPRInfo::callFrameRegister);
+        m_jit.storePtr(CCallHelpers::TrustedImmPtr(nullptr), CCallHelpers::Address(GPRInfo::regT0, VM::callFrameForCatchOffset()));
+
+        JIT_COMMENT(m_jit, "Configure wasm context instance");
+        m_jit.loadPtr(CCallHelpers::Address(GPRInfo::callFrameRegister, CallFrameSlot::codeBlock * sizeof(Register)), GPRInfo::wasmContextInstancePointer);
+        m_frameSizeLabels.append(m_jit.moveWithPatch(TrustedImmPtr(nullptr), GPRInfo::regT0));
+        m_jit.subPtr(GPRInfo::callFrameRegister, GPRInfo::regT0, MacroAssembler::stackPointerRegister);
+    }
+
+    void emitCatchAllImpl(ControlData& dataCatch)
+    {
+        m_catchEntrypoints.append(m_jit.label());
+        emitCatchPrologue();
+        restoreWebAssemblyGlobalState();
+        m_jit.prepareWasmCallOperation(GPRInfo::wasmContextInstancePointer);
+        m_jit.setupArguments<decltype(operationWasmRetrieveAndClearExceptionIfCatchable)>(GPRInfo::wasmContextInstancePointer);
+        m_jit.callOperation(operationWasmRetrieveAndClearExceptionIfCatchable);
+        static_assert(noOverlap(GPRInfo::nonPreservedNonArgumentGPR0, GPRInfo::returnValueGPR, GPRInfo::returnValueGPR2));
+        bind(this->exception(dataCatch), Location::fromGPR(GPRInfo::returnValueGPR));
+        Stack emptyStack { };
+        dataCatch.startBlock(*this, emptyStack);
```

Comment:
That sounds good. I would like to keep

```
        bind(this->exception(dataCatch), Location::fromGPR(GPRInfo::returnValueGPR));
        Stack emptyStack { };
        dataCatch.startBlock(*this, emptyStack);
```

separately since they are more tied to BBQJIT's mechanism (not just a JIT code generator).